### DISCUSSION TITLE
Inverter Omnik 3000TL not connecting

### DISCUSSION THREAD

**Discussion by rulaan** [robbinjanssen/home-assistant-omnik-inverter#134] [Q&A]
Hi,

My 2nd omnik inverter is not connecting to HA (the first one, an Omniksol-2.5k-TL3-S bought in 2019 connected on first attempt). I tried the 4 options provided, and also tried the URL's mentioned.
JS => 404
JSON => 404
HTML => 404

The converter is an Omniksol 3000 TL bought in 2013. Has been connected to omnikportal, and is now connected to solarmanpv (so there is output to be read).

I am able to access the following page on the logger (after authenticating with a changed admin user and password):
http://<inverter_IP>/home.html

Any suggestions on making the connection work are welcome!

Thanks, Rudi.

**Comment by klaasnicolaas** [comment]
Have you also tried the TCP option?

**Comment by rulaan** [comment]
Yes

**Comment by robbinjanssen** [comment]
Hi @rulaan!

1. What happens when you access http://<inverter_IP>/js/status.js ?

2. Can you check your HTML source and see if there is any JS data? It should contain something like: https://github.com/klaasnicolaas/python-omnikinverter/blob/main/tests/fixtures/status.html

**Comment by rulaan** [comment]
Hi Robbinjanssen,

Trying to connect with JS/JSON/HTML gives:
![image](https://user-images.githubusercontent.com/18302947/192348660-09c80c06-9ce5-43cc-9986-3868dd519c9b.png)
![image](https://user-images.githubusercontent.com/18302947/192348821-c4af8be1-d3fc-489f-9844-34b1ed981e8a.png)


Trying with TCP gives (note that a leading zero for the serial is stripped,
should be: 0602321995):
![image](https://user-images.githubusercontent.com/18302947/192349063-13a9804d-a754-4d6f-89e4-42ced03787da.png)


Testing  http://<inverter_IP>/js/status.js  on the Omnik-2.5k gives a response as expected.

Testing  http://<inverter_IP>/js/status.js on the Omnik-3k gives a 404 Not Found. No more info hidden in the html source.

Maybe the logger is too old to have more than just a landing page
(../home.html)?

Thanks for helping.

**Comment by robbinjanssen** [comment]
Hi @rulaan, unfortunately your images don't work. Can you try uploading them on github directly?

**Comment by rulaan** [comment]
> Hi @rulaan, unfortunately your images don't work. Can you try uploading them on github directly?

Added in post above.

**Comment by htelgen** [comment]
I have the same problem. My other converter Omnik 1k is still connected

**Comment by robbinjanssen** [comment]
I'm afraid that if js/html doesn't work the only option is TCP. But that's kinda hard to debug (especially remotely). @MarijnS95 any idea on how we can get some debug info on this?

**Comment by MarijnS95** [comment]
@robbinjanssen I thought we looked into improving both the logging and status messages in the UI when currently "Can't connect" or "Unknown error" is displayed, but don't remember if that landed.

For starters, @rulaan @htelgen what appears in the Home Assistant logs right as `Unknown error occurred` is shown when attempting a connection through the TCP backend?

If that doesn't surface anything useful I'll recheck where the debug code went (and PR it back in), but we can always defer to something like [a filtered WireShark network capture](https://github.com/robbinjanssen/home-assistant-omnik-inverter/issues/121#issuecomment-1208217981) to inspect the exchanged data and go from there.

**Comment by remcovl92** [comment]
I have the same problem with Omniksol 3000TL. 
The data logger has firmware version 4.01.9.d1, maybe this is not supported in this firmware version?
@rulaan & @htelgen what firmware do you have?

**Comment by htelgen** [comment]
Where would I find that information?

Op za 15 okt. 2022 om 17:14 schreef Remco Vliegen ***@***.***

**Comment by remcovl92** [comment]
On the home page of the logger http://<inverter_IP>/home.html
then at the bottom of the page:
![image](https://user-images.githubusercontent.com/47039263/196031791-a771c741-f9ec-45fb-9d63-daa40829dcf7.png)

**Comment by htelgen** [comment]
My problem, at the moment, is that both my converters are not accessible
via Wifi anymore, so I don't get to the information.
I will try to reset the inverters (by cutting off the power) and try
accessing them again.

Op zo 16 okt. 2022 om 13:06 schreef Remco Vliegen ***@***.***

**Comment by robbinjanssen** [comment]
@remcovl92 @htelgen I doubt the firmware version matters. To properly assist you we need some more debugging logs like @MarijnS95 suggested in https://github.com/robbinjanssen/home-assistant-omnik-inverter/issues/129#issuecomment-1273093698 :-)

**Comment by htelgen** [comment]
I was able to connect to my 3000tl converter. It has ver:1.0.07
Any idea where to get an upgrade?

**Comment by robbinjanssen** [comment]
Converting this into a discussion as it is turning into a support thread ;-) Feel free to help each other out!

**Comment by Pluimvee** [comment]
Hi All,

Great to see so many people putting so much effort in having Omnik invertors connected to AH. Thanks for all the work.

Unfortunate I have the same problem as mentioned above: an older Omnik 2000TL which does not support status.js/html/json. Also the TCP option doen't connect.

I noticed when the inverter is running in AP mode there is another webpage compared to when the invertor is running in STA mode. In AP mode I see a blue screen with version 1.0.04, in STA mode I end up on a 'M2M Server' webpage (of the data logger?) which has a version 3.29.10He.

When looking for firmware upgrades, RS485 boards, ESP proxy and alternative solutions I saw the option to configure multiple remote servers in the AP-mode configuration screen (is this the invertor config page?)
To try this option I created a small webserver running on an ESP8266 which forwards the TCP payload to serial and to mosquitto MQTT which I can read into HA. I configured the ip address of the ESP into the invertor and WHALA! I have an old omnik connected to HA.

However: running an ESP just to have the payload forwarded to MQTT and then to HA is many hops, with complexity, overkill, maintanance, availability issues, etc. So I'm trying to get some modules removed in between.

The best option is to have the Omnik invertor post directly to HA. Meaning: HA will not poll the Omnik, Omnik will post to HA. Since I'm well acquinted with C++ but have no knowledge at all on Python I need some help. 

- Is it possible to have a (or this) modeul to listen on a TCP port to receive postings from the Omnik invertor?
- Would it be better to have mosquitto in between as MQTT is exactly meant to buffer telemetry, and would it then be better to build a seperate component, not running in HA but as add-on in a seperate docker container?

Please let me know your thoughts

The STA-mode webpage
<img width="445" alt="Screenshot 2023-03-16 163304" src="https://user-images.githubusercontent.com/124380379/225671729-2b11226d-25f8-4890-84b3-1ab40d37096e.png">

The AP-mode webpage
<img width="734" alt="Screenshot 2023-03-16 090611" src="https://user-images.githubusercontent.com/124380379/225672924-a9f2de64-1293-4516-bb9a-6bdf06de0a06.png">

**Reply by Pluimvee** [reply]
Ps I asume the TCP option uses the polling on port 8899, wich my logger doesnt listen on. So even that option my invertor/logger doesnt support

**Reply by robbinjanssen** [reply]
I don't think its possible for an integration to setup an "Endpoint" to which your omnik can connect. I think it's possible using node-red though? Maybe use node red and then transform it to mqtt (sensors)? Unfortunately this package won't help you with this situation

**Reply by Pluimvee** [reply]
Thanks robbin,
I started a fork of the OmnikLoggerProxy of @jbouwh -> getting myself acquinted w/ github and python. Its a steep learning curve but I got it working this weekend using a python script, receiving UDP packages, parse the package and post the sensors to MQTT. 
Still needs some work to get the individual sensors and possibly to have it running in AppDeamon

**Comment by Sinta77** [comment]
Hi all, first thank you all for the great work. Always much appreciated.. I am new to HA and just set it up and have it running and now i'm trying to connect it to my omnik 3k-tl2 invertor. Tried all the options and the only one with a small succes is the TCP one. But that is giving me an error that the buffer size is too small:  Buffer size too small (85 instead of at least 125 bytes)

Any ideas on how to fix this?

**Reply by jbouwh** [reply]
You could try to remove the size check (it is hard coded to 99 bytes). If that works I could update the code.

**Reply by MarijnS95** [reply]
This isn't an arbitrary hardcoded limit you can easily adjust, it's the struct layout that parses known fields. There was another inverter missing exactly the last 40 bytes (https://github.com/robbinjanssen/home-assistant-omnik-inverter/issues/121#issuecomment-1208734235) and those contain exactly the firmware strings.

I should be able to solve both by making the firmware string optional, but am also really curious if we can use any of the other bytes to validate the type/version of the message. See also how more constants had to be changed to make it work, and the inverter responding with different data in some of the "yet unknown meaning" bytes.

Perhaps you can share a data dump of your inverter, i.e. via wireshark capture? Or at the very least any logging that comes out of this plugin (and the underlying `python-omnikinverter` package) before running into the error?

**Reply by MarijnS95** [reply]
@jbouwh where'd you find this hardcoded value of `99`, perhaps in your own data logger? That is in fact the data size in one TCP packet including the start and end markers, message length and checksum, and size. After parsing those ([see our TCP implementation](https://github.com/klaasnicolaas/python-omnikinverter/blob/main/omnikinverter/tcp.py)) to extract the inner data we are left with these 85 bytes (plus 1 start and 1 end byte, plus 12 header bytes, equals 99!) of useful inverter info/statistics.

Coincidentally we already have an issue open about this (and it's not the Trannergy issue, that seems to have a similar internal data layout but more bytes are different) and @PeeVv shared a 3000TL dump today at https://github.com/robbinjanssen/home-assistant-omnik-inverter/issues/142#issuecomment-1507430793 that made me aware of the sizes, and confirm the message contents to write a solution. Many thanks!

@Sinta77 let's move further discussion on this concrete parser error to https://github.com/robbinjanssen/home-assistant-omnik-inverter/issues/142, leaving this discussion board for connection issues. But I hope we can close it soon after I have submitted a PR for this.
